### PR TITLE
JAMES-2721 reduce test time in mailbox/cassandra

### PR DESCRIPTION
**Pull request by Arsnael**
The goal here is to reduce the testing time of `mailbox/cassandra` module by re-enabling the forks and using a new restart strategy for the Cassandra docker container.

We increment a counter when we run a test and when it's more than the max defined (we start with 500 here, might change depending on how well the CI can support it), we restart the container on the next test class instance, before starting the cluster. We used some hacks (specially for JUNIT5 extension) to keep the code as composed as possible. 

Don't hesitate to give any feedback on this if you think having a better approach !

Local tests execution time : 
* before refactoring : ~ 21 minutes
* after refactoring : ~ 3 - 4 minutes

### CODE REVIEW

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/host/CassandraRestartExtension.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ ****************************************************************/
+
+package org.apache.james.mailbox.cassandra.host;
+
+import org.apache.james.backends.cassandra.DockerCassandraSingleton;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+public class CassandraRestartExtension implements BeforeAllCallback, BeforeEachCallback {
```
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/host/CassandraRestartExtension.java]: This class should be in backand/cassandra IMO

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/host/DockerCassandraIncrementTestsPlayedRule.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.cassandra.host;
+
+import org.apache.james.backends.cassandra.DockerCassandraSingleton;
+import org.junit.rules.ExternalResource;
+
+public class DockerCassandraIncrementTestsPlayedRule extends ExternalResource {
```
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/host/DockerCassandraIncrementTestsPlayedRule.java]: Idem, this should be in backend/cassandra

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/host/DockerCassandraRestartRule.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+package org.apache.james.mailbox.cassandra.host;
+
+import org.apache.james.backends.cassandra.DockerCassandraSingleton;
+import org.junit.rules.ExternalResource;
+
+public class DockerCassandraRestartRule extends ExternalResource {
+    @Override
+    protected void before() {
+        DockerCassandraSingleton.restartAfterMaxTestsPlayed();
```
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/host/DockerCassandraRestartRule.java]: Idem, this should be in backend/cassandra

**Comment by Arsnael** [comment]
```
[1315c59850f0d709531636381054081b6d891f18] [ERROR] Failures: 
[1315c59850f0d709531636381054081b6d891f18] [ERROR]   CassandraMessageIdToImapUidDAOTest.updateShouldUpdateFlaggedFlag:303 
[1315c59850f0d709531636381054081b6d891f18] Expecting:
[1315c59850f0d709531636381054081b6d891f18]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=4bb83601-55c6-11e9-ace4-ff014616218e}, messageId=CassandraMessageId{uuid=4bb83600-55c6-11e9-ace4-ff014616218e}, uid=MessageUid{uid=1}}, flags=, modSeq=1}]>
[1315c59850f0d709531636381054081b6d891f18] to contain only:
[1315c59850f0d709531636381054081b6d891f18]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=4bb83601-55c6-11e9-ace4-ff014616218e}, messageId=CassandraMessageId{uuid=4bb83600-55c6-11e9-ace4-ff014616218e}, uid=MessageUid{uid=1}}, flags=\Flagged, modSeq=2}]>
[1315c59850f0d709531636381054081b6d891f18] elements not found:
[1315c59850f0d709531636381054081b6d891f18]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=4bb83601-55c6-11e9-ace4-ff014616218e}, messageId=CassandraMessageId{uuid=4bb83600-55c6-11e9-ace4-ff014616218e}, uid=MessageUid{uid=1}}, flags=\Flagged, modSeq=2}]>
[1315c59850f0d709531636381054081b6d891f18] and elements not expected:
[1315c59850f0d709531636381054081b6d891f18]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=4bb83601-55c6-11e9-ace4-ff014616218e}, messageId=CassandraMessageId{uuid=4bb83600-55c6-11e9-ace4-ff014616218e}, uid=MessageUid{uid=1}}, flags=, modSeq=1}]>
[1315c59850f0d709531636381054081b6d891f18] 
[1315c59850f0d709531636381054081b6d891f18] [ERROR]   CassandraMessageIdToImapUidDAOTest.updateShouldUpdateRecentFlag:328 
[1315c59850f0d709531636381054081b6d891f18] Expecting:
[1315c59850f0d709531636381054081b6d891f18]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=4ba7e251-55c6-11e9-ace4-ff014616218e}, messageId=CassandraMessageId{uuid=4ba7e250-55c6-11e9-ace4-ff014616218e}, uid=MessageUid{uid=1}}, flags=, modSeq=1}]>
[1315c59850f0d709531636381054081b6d891f18] to contain only:
[1315c59850f0d709531636381054081b6d891f18]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=4ba7e251-55c6-11e9-ace4-ff014616218e}, messageId=CassandraMessageId{uuid=4ba7e250-55c6-11e9-ace4-ff014616218e}, uid=MessageUid{uid=1}}, flags=\Recent, modSeq=2}]>
[1315c59850f0d709531636381054081b6d891f18] elements not found:
[1315c59850f0d709531636381054081b6d891f18]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=4ba7e251-55c6-11e9-ace4-ff014616218e}, messageId=CassandraMessageId{uuid=4ba7e250-55c6-11e9-ace4-ff014616218e}, uid=MessageUid{uid=1}}, flags=\Recent, modSeq=2}]>
[1315c59850f0d709531636381054081b6d891f18] and elements not expected:
[1315c59850f0d709531636381054081b6d891f18]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=4ba7e251-55c6-11e9-ace4-ff014616218e}, messageId=CassandraMessageId{uuid=4ba7e250-55c6-11e9-ace4-ff014616218e}, uid=MessageUid{uid=1}}, flags=, modSeq=1}]>
[1315c59850f0d709531636381054081b6d891f18] 
[1315c59850f0d709531636381054081b6d891f18] [ERROR]   CassandraMessageIdToImapUidDAOTest.updateShouldUpdateSeenFlag:354 
[1315c59850f0d709531636381054081b6d891f18] Expecting:
[1315c59850f0d709531636381054081b6d891f18]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=4bbf61f1-55c6-11e9-ace4-ff014616218e}, messageId=CassandraMessageId{uuid=4bbf61f0-55c6-11e9-ace4-ff014616218e}, uid=MessageUid{uid=1}}, flags=, modSeq=1}]>
[1315c59850f0d709531636381054081b6d891f18] to contain only:
[1315c59850f0d709531636381054081b6d891f18]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=4bbf61f1-55c6-11e9-ace4-ff014616218e}, messageId=CassandraMessageId{uuid=4bbf61f0-55c6-11e9-ace4-ff014616218e}, uid=MessageUid{uid=1}}, flags=\Seen, modSeq=2}]>
[1315c59850f0d709531636381054081b6d891f18] elements not found:
[1315c59850f0d709531636381054081b6d891f18]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=4bbf61f1-55c6-11e9-ace4-ff014616218e}, messageId=CassandraMessageId{uuid=4bbf61f0-55c6-11e9-ace4-ff014616218e}, uid=MessageUid{uid=1}}, flags=\Seen, modSeq=2}]>
[1315c59850f0d709531636381054081b6d891f18] and elements not expected:
[1315c59850f0d709531636381054081b6d891f18]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=4bbf61f1-55c6-11e9-ace4-ff014616218e}, messageId=CassandraMessageId{uuid=4bbf61f0-55c6-11e9-ace4-ff014616218e}, uid=MessageUid{uid=1}}, flags=, modSeq=1}]>
[1315c59850f0d709531636381054081b6d891f18] 
[1315c59850f0d709531636381054081b6d891f18] [ERROR]   CassandraMessageIdToImapUidDAOTest.updateShouldUpdateUserFlag:379 
[1315c59850f0d709531636381054081b6d891f18] Expecting:
[1315c59850f0d709531636381054081b6d891f18]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=4bafd190-55c6-11e9-ace4-ff014616218e}, messageId=CassandraMessageId{uuid=4bafaa80-55c6-11e9-ace4-ff014616218e}, uid=MessageUid{uid=1}}, flags=, modSeq=1}]>
[1315c59850f0d709531636381054081b6d891f18] to contain only:
[1315c59850f0d709531636381054081b6d891f18]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=4bafd190-55c6-11e9-ace4-ff014616218e}, messageId=CassandraMessageId{uuid=4bafaa80-55c6-11e9-ace4-ff014616218e}, uid=MessageUid{uid=1}}, flags=\*, modSeq=2}]>
[1315c59850f0d709531636381054081b6d891f18] elements not found:
[1315c59850f0d709531636381054081b6d891f18]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=4bafd190-55c6-11e9-ace4-ff014616218e}, messageId=CassandraMessageId{uuid=4bafaa80-55c6-11e9-ace4-ff014616218e}, uid=MessageUid{uid=1}}, flags=\*, modSeq=2}]>
[1315c59850f0d709531636381054081b6d891f18] and elements not expected:
[1315c59850f0d709531636381054081b6d891f18]   <[ComposedMessageIdWithMetaData{composedMessageId=ComposedMessageId{mailboxId=CassandraId{id=4bafd190-55c6-11e9-ace4-ff014616218e}, messageId=CassandraMessageId{uuid=4bafaa80-55c6-11e9-ace4-ff014616218e}, uid=MessageUid{uid=1}}, flags=, modSeq=1}]>
```

**Comment by Arsnael** [comment]
The issue above is something I remember encountering before and that I could reproduce randomly on my local when running many times the build for this module. It could be that `500` tests was too much before doing a container restart. I just switched it to `200`. It seems I cannot reproduce this issue locally anymore now.

We have 790 tests in this module, so with 200 it means we restart 3 times, which is still acceptable.

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxPathDAOTest.java [hunk 4]
```diff
@@ -27,26 +27,29 @@
...
+    private static final String USER = "user";
+    private static final String OTHER_USER = "other";
+
+    static final CassandraId INBOX_ID = CassandraId.timeBased();
+    static final CassandraId OUTBOX_ID = CassandraId.timeBased();
+    static final CassandraId otherMailboxId = CassandraId.timeBased();
+    static final MailboxPath USER_INBOX_MAILBOXPATH = MailboxPath.forUser(USER, "INBOX");
+    static final MailboxPath USER_OUTBOX_MAILBOXPATH = MailboxPath.forUser(USER, "OUTBOX");
+    static final MailboxPath OTHER_USER_MAILBOXPATH = MailboxPath.forUser(OTHER_USER, "INBOX");
+    static final CassandraIdAndPath INBOX_ID_AND_PATH = new CassandraIdAndPath(INBOX_ID, USER_INBOX_MAILBOXPATH);
```
Comment (rouazana) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxPathDAOTest.java]: please avoid unrelated changes, especially on big commits

Comment (Arsnael) [reply mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxPathDAOTest.java]: Ok, I thought it was ok to change the code to fix a bit more but I understand and will try to revert those. Thanks for feedback !

Comment (rouazana) [reply mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxPathDAOTest.java]: no problem to fix a little more, just do it in a separated commit

File: backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/DockerCassandraSingleton.java [hunk 5]
```diff
@@ -19,15 +19,32 @@
...
 
     public static final DockerCassandra singleton = new DockerCassandra();
 
     static {
         singleton.start();
     }
 
-    public static void restart() {
+    public static void incrementTestsPlayed() {
+        testsPlayedCount.incrementAndGet();
```
Comment (mbaechler) [inline backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/DockerCassandraSingleton.java]: what's the point of using an AtomicInteger as the restart is obviously not thread-safe

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/host/CassandraRestartExtension.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Override
+    public void beforeAll(ExtensionContext extensionContext) {
+        DockerCassandraSingleton.restartAfterMaxTestsPlayed();
```
Comment (mbaechler) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/host/CassandraRestartExtension.java]: Why it's not in beforeEach?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/host/DockerCassandraIncrementTestsPlayedRule.java (hunk 2 above)
Comment (mbaechler) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/host/DockerCassandraIncrementTestsPlayedRule.java]: Why would we want to count for tests and never restart cassandra?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/CassandraCombinationManagerTest.java [hunk 7]
```diff
@@ -33,11 +35,17 @@
 import org.junit.Before;
 import org.junit.BeforeClass;
 import org.junit.ClassRule;
+import org.junit.Rule;
 
 public class CassandraCombinationManagerTest extends AbstractCombinationManagerTest {
 
     @ClassRule public static DockerCassandraRule cassandraServer = new DockerCassandraRule();
 
+    @ClassRule public static DockerCassandraRestartRule cassandraRestartRule = new DockerCassandraRestartRule();
```
Comment (mbaechler) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/CassandraCombinationManagerTest.java]: please have a look at https://github.com/linagora/james-project/pull/941/files#diff-c9bcff730d848155c53692dd5c9349e8R29 : I solved that problem more elegantly by providing a static factory that allows to configure the Cassandra container lifecycle

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/host/CassandraRestartExtension.java (hunk 6 above)
Comment (Arsnael) [reply mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/host/CassandraRestartExtension.java]: Tried at first, it is not working. It seems (if I understood well) that if you try to restart the container after creating and starting the cluster, it crashes the rest of your test suite

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/host/DockerCassandraIncrementTestsPlayedRule.java (hunk 2 above)
Comment (Arsnael) [reply mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/host/DockerCassandraIncrementTestsPlayedRule.java]: That's in the other rule, as I was trying to restart in a beforeClass manner. Will restudy the question with your new inputs

**Comment by hoangdat** [comment]
Read it

**Comment by Arsnael** [comment]
> That's a bit unfortunate that you were not tell we already implemented a similar solution in the past.
> Anyway, maybe you can merge both solutions for the best

I didn't have that much time to look deeply into what has been done before but as I understand, your precedent solution seems to create a new `ContainerLifecycleConfiguration`, with a new counter, etc. So you do restarts based on the number of tests in a test class. Here we are trying to keep track of tests globally on the module.

So I still tried to use it but it didn't work. I encountered the same issue as I had with my rules, as if I do a restart in the middle of a test class, after the cassandra cluster started, it seems not happy of loosing the connexion to the cassandra container and crashes for the rest of the test suite. 

So if you have an idea I'm all ears, if not I will try to look more into it later.

**Comment by chibenwa** [comment]
test this please

**Comment by mbaechler** [comment]
> @mbaechler : I didn't have that much time to look deeply into what has been done before but as I understand, your precedent solution seems to create a new ContainerLifecycleConfiguration, with a new counter, etc. for every test class implementing it. So you do restarts based on the number of tests in a test class. Here we are trying to keep track of tests globally on the module.

It's per-class. BTW, the singleton strategy is newer than my previous solution.
However, my comment was not about the implementation but more about how to use the extension/rule, you can very well use the builder pattern and keep your implementation.

> So I still tried to use it but it didn't work. I encountered the same issue as I had with my rules, as if I do a restart in the middle of a test class, after the cassandra cluster started, it seems not happy of loosing the connexion to the cassandra container and crashes for the rest of the test suite

I'm very worried about that: you start a new james server for every single test, why restarting cassandra could have any impact?

**Comment by chibenwa** [comment]
> I'm very worried about that: you start a new james server for every single test, why restarting cassandra could have any impact?

We start James server within the `mailbox/cassandra` project now?

**Comment by mbaechler** [comment]
> We start James server within the mailbox/cassandra project now?

No, of course.

It's even more worrying, isn't it?

**Comment by Arsnael** [comment]
At the end I only change the Rule, to restart cassandra potentially between tests in those classes. I tried to check to do the same with the Extension, but other modules and tests are reusing the `CassandraClusterExtension` and it would require some refactoring as well there. I didn't want to impact too much outside of `mailbox/cassandra` in this PR. If you have an opinion about that let me know !

**Comment by chibenwa** [comment]
Merged
